### DISCUSSION TITLE
KeepLargestConnectedComponent

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#2625] [General]
Hi, 

I was using the KeepLargestConnectedComponent as a post transform for segmentation tasks and it always worked well, since two weeks it shows me a new error saying: 
NotImplementedError: Only for images of dimension 1-3 are supported, got a 4D one

I am wondering why, I did not change the dimensions of my data.
here is my notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

thanks for your help

**Comment by Nic-Ma** [comment]
Hi @Meddebma ,

Thanks for your interest and experiments with MONAI.
We introduced some minor breaking change in v0.6 recently, please check this doc for more details:
https://github.com/Project-MONAI/MONAI/wiki/v0.5-to-v0.6-migration-guide.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma 
Thanks for the information, I cant see the implementation of KeepLargestConnectedComponent in any notebook now unfortunately.. 

Hi @rijobro,
do you think it has something to do with the changes in v0.6? how should I implement KeepLargestConnectedComponent in the post-trans?
thanks

**Reply by rijobro** [reply]
Hi @Meddebma do you think you could create a minimum working example that demonstrates the error?

**Reply by rijobro** [reply]
@Nic-Ma I just had a look at the source code and admit that I find it confusing too. From the docs, it says it's a post-transform, but also says that it wants shape `C, H, W, [D]`, i.e., it doesn't accept the batch dimension. So I'm not sure how it could work as a post-transform: https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L185-L189.

**Reply by Meddebma** [reply]
Hi @rijobro, yeah sure, here is the notebook:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

**Reply by Nic-Ma** [reply]
Hi @rijobro ,

As we discussed, now we decollate a batch and execute postprocessing transforms for every item of the decollated list.
So it should be `channel-first` transform instead of `batch-first` transform, this doc shows more details:
https://github.com/Project-MONAI/MONAI/wiki/v0.5-to-v0.6-migration-guide

Thanks.

**Reply by rijobro** [reply]
of course, thanks @Nic-Ma!

**Reply by Meddebma** [reply]
@rijobro
did you have the opportunity to look at the notebook what should I change to use the keeplargestcomponent?
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb
thank you

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

As you are using MONAI v0.6, please change the postprocessing logic to decollate batch first, then execute transforms for every channel-first item.
```py
from monai.data import decollate_batch

val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
```

Thanks.

**Reply by Meddebma** [reply]
Hi @rijobro 
thanks for your help. The postprocessing worked well. I still have a problem with the evaluation,
I would like to display for every data the segmentation mask as well as segmentation metrics, 

for i, val_data in enumerate(val_loader):
        val_images, val_labels = val_data["image"].to(device), val_data["label"].to(device)

TypeError: new(): invalid data type 'str'

here is the notebook
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

**Reply by Nic-Ma** [reply]
I feel maybe you didn't execute transforms for `image` and `label`, so `val_data["image"]` and `val_data["label"]` are still file path strings, maybe you can print out to debug?

Thanks.

**Reply by Meddebma** [reply]
when I write
"for val_data in val_loader "
it works properly, 
I did a loop into another and I get this result:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb
